src/libertem/common/buffers.py

```diff
@@ -167,6 +180,141 @@ def __del__(self):
         self.pool.checkin_bytes(self.size, self.alignment, self.buf)
 
 
+T = TypeVar('T', bound=np.generic)
+
+
+class InvalidMaskError(Exception):
+    """
+    The mask is not compatible, raised for example when the shape of
+    the mask doesn't match the data.
+    """
+    pass
+
+
+class ArrayWithMask(Generic[T]):
+    """
+    An opaque type representing an array together with a
+    mask (for example, defining a region in the array
+    where there are valid entries)
+
+    This is meant for usage in :meth:`UDF.get_results`,
+    and you can use the convenience method :meth:`UDF.with_mask`
+    to instantiate it.
+    """
+
+    def __init__(self, arr: T, mask: Union[np.ndarray, bool]):
+        if isinstance(mask, int):
+            mask = np.array([mask])
```

Comment:
The signature says `ndarray` or `bool`, and in the next line we test for `int`, which is a superset of `bool`. Probably the parameters should be documented in the docstring?

* `ndarray` of what types?
* What is the difference between dtype `int` and `bool` ndarrays?

The behavior of NumPy fancy indexing is different with an `int` mask or `bool` mask: With an `int` mask it returns the elements at those indices. I can't quite imagine how that should work with valid data masking, so we should probably restrict the mask to dtype `bool`?

Reply:
True, this was confusing, the type should be `bool` or `np.ndarray`. I'll add documentation to `with_mask`, which is the public interface, to clear this up, and I'll consistently use `bool` instead of `int` (for Python this doesn't really matter, and I somehow got into the habit of using `1` instead of `True` sometimes, but it's clear that there's a difference for `numpy`...). The intention of the interface was to pass in boolean masks, not lists of coordinates, which could be added as a follow-up IMHO.

Reply:
Yeah, list of coordinates will also modify the shape and contain duplicates, so I think bool only makes sense for now. :-)